sympy/polys/matrices/dense.py

```diff
@@ -160,8 +212,229 @@ def ddm_irref(a, _partial_pivot=False):
+    [0, 1]
+    >>> Matrix(M).rref()[0]
+    Matrix([
+    [1, 0, -1],
+    [0, 1,  2]])
+
+    See Also
+    ========
+
+    ddm_irref
+        A version of this routine that uses field division.
+    sdm_irref
+        A sparse version of :func:`ddm_irref`.
+    sympy.polys.matrices.domainmatrix.DomainMatrix.rref_den
+        Higher level interface.
+
+    References
+    ==========
+
+    .. [1] Fraction-free algorithms for linear and polynomial equations.
+        George C. Nakos , Peter R. Turner , Robert M. Williams.
+        https://dl.acm.org/doi/10.1145/271130.271133
+    """
+    #
+    # A simpler presentation of this algorithm is given in [1]:
+    #
+    # Given an n x n matrix A and n x 1 matrix b:
+    #
+    #   for i in range(n):
+    #       if i != 0:
+    #           d = a[i-1][i-1]
+    #       for j in range(n):
+    #           if j == i:
+    #               continue
+    #           b[j] = a[i][i]*b[j] - a[j][i]*b[i]
+    #           for k in range(n):
+    #               a[j][k] = a[i][i]*a[j][k] - a[j][i]*a[i][k]
+    #               if i != 0:
+    #                   a[j][k] /= d
+    #
+    # Our version here is a bit more complicated because:
+    #
+    #  1. We use row-swaps to avoid zero pivots.
+    #  2. We allow for some columns to be missing pivots.
+    #  3. We avoid a lot of redundant arithmetic.
+    #
+    # TODO: Use a non-trivial pivoting strategy. Even just row swapping makes a
+    # big difference to performance if e.g. the upper-left entry of the matrix
+    # is a huge polynomial.
+
+    # a is (m x n)
+    m = len(a)
+    if not m:
+        return K.one, []
+    n = len(a[0])
+
+    d = None
+    pivots = []
+    no_pivots = []
+
+    # i, j will be the row and column indices of the current pivot
+    i = 0
+    for j in range(n):
+        # next pivot?
+        aij = a[i][j]
+
+        # swap rows if zero
+        if not aij:
+            for ip in range(i+1, m):
+                aij = a[ip][j]
+                # row-swap
+                if aij:
+                    a[i], a[ip] = a[ip], a[i]
+                    break
+            else:
+                # go to next column
+                no_pivots.append(j)
+                continue
+
+        # Now aij is the pivot and i,j are the row and column. We need to clear
+        # the column above and below but we also need to keep track of the
+        # denominator of the RREF which means also multiplying everything above
+        # and to the left by the current pivot aij. This upper left corner is
+        # usually already diagonal with all diagonal entries equal but there
+        # can be other non-zero entries in any column that has no pivot.
+        if not K.is_one(aij):
+
+            # Update previous pivots in the matrix
+            if pivots:
+                pivot_val = aij * a[0][pivots[0]]
+                # Divide out the common factor
+                if d is not None:
+                    pivot_val = K.exquo(pivot_val, d)
```

Comment:
If `aij` is one then we skip multiplying the upper left corner. We also skip dividing by `d` though but `d` might not be one.

This might not be correct...

Reply:
Here `d` is the previous pivot and `aij` is the current pivot. If we are working over e.g. `ZZ` then each iteration that multiples `aij` would carry that multiplication over to subsequent iterations. Maybe over `QQ`  though it is possible that `aij` could be one even if `d` is not...

Reply:
I can't find an example that breaks this. Maybe the code is fine but I'm just too tired right now to see why it is fine (I must have thought that it was fine when I wrote it...).

Reply:
This was wrong and I found a counter example:
```python
In [1]: M = Matrix.diag([S(1)/2, 2])

In [2]: M.to_DM().rref_den()[0].to_Matrix()
Out[2]: 
⎡1/2  0⎤
⎢      ⎥
⎣ 0   1⎦

In [3]: M.rref()
Out[3]: 
⎛⎡1  0⎤        ⎞
⎜⎢    ⎥, (0, 1)⎟
⎝⎣0  1⎦        ⎠
```
The proper check would be
```python
if not (K.is_one(aij) and d is None):
```
I decided that this optimisation is not worth having though. The pivot and denominator will rarely both be `1` and this would only saving a handful of operations like `1 * a` and `exquo(a, 1)` which should be cheap operations any way.

I have just removed the `if` check so this is fixed.